Python/Product/Cookiecutter/Model/CookiecutterClient.cs

```diff
@@ -30,33 +30,38 @@ namespace Microsoft.CookiecutterTools.Model {
     class CookiecutterClient : ICookiecutterClient {
         private readonly IServiceProvider _provider;
         private readonly CookiecutterPythonInterpreter _interpreter;
-        private readonly string _envFolderPath;
-        private readonly string _envInterpreterPath;
+        private readonly string _expectedEnvFolderPath;
+        private string _envFolderPath;
+        private string _envInterpreterPath;
         private readonly Redirector _redirector;
 
         internal string DefaultBasePath { get; set; } = Environment.GetFolderPath(Environment.SpecialFolder.MyDocuments);
 
+
         public bool CookiecutterInstalled {
             get {
-                if (!File.Exists(_envInterpreterPath)) {
-                    return false;
-                }
-
-                return true;
+                return File.Exists(_envInterpreterPath);
             }
         }
 
         public CookiecutterClient(IServiceProvider provider, CookiecutterPythonInterpreter interpreter, Redirector redirector) {
             _provider = provider;
             _interpreter = interpreter;
             var localAppDataFolderPath = Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData);
-            _envFolderPath = Path.Combine(localAppDataFolderPath, "Microsoft", "CookiecutterTools", "env");
-            _envInterpreterPath = Path.Combine(_envFolderPath, "scripts", "python.exe");
+
+            // This is where the env *should* be created, but not necessarily where it will live on disk.
+            // See the GetRealPath function for more details
+            _expectedEnvFolderPath = Path.Combine(localAppDataFolderPath, "Microsoft", "CookiecutterTools", "env");
+
+            // get the real paths to the env and interpreter, in case they've been redirected
+            _envFolderPath = Task.Run(() => GetRealPath(_expectedEnvFolderPath)).Result;
```

Comment:
Using `.Result` on a Task is equivalent to just doing the operation directly (except it's more likely to cause a deadlock).

To do this properly, you'd start the task running and keep a reference to it on the instance, then `await` it to get the result when it's actually needed.

Reply:
@zooba I did what you suggested. It's a bit of code duplication, but I don't see how I can avoid that. Please let me know if this is what you had in mind.

Thanks!

Reply:
I was able to simplify this logic a bit. The _expectedEnvFolderPath is the only path supplied to the constructor, so instead of having a common task, all the methods just check to see if _envFolderPath has been set already. If it has, we've already checked for redirection, and it's a no-op.

The only exception is the method that creates a new venv, in which case it needs to check for redirection again, even if _envFolderPath is set from before. This is done via a boolean that forces the redirection check (python call) to run again.